### PR TITLE
refactor: extract shared row-partitioning helpers

### LINKED ISSUE #60: refactor: extract shared entity-row split/reorder/recombine helper
## Context

Three workflows now implement the same pattern:
1. Add a `_anonymizer_row_order` column to preserve original ordering.
2. Partition DataFrame into entity rows vs passthrough rows.
3. Run LLM/column pipeline on entity rows only.
4. Set a default value on passthrough rows (e.g., empty replacement map, original text).
5. `pd.concat` + `sort_values("_anonymizer_row_order")` + `drop(columns=...)` + `reset_index`.
6. Propagate `df.attrs`.

**Locations**:
- `engine/replace/llm_replace_workflow.py` — `generate_map_only()`.
- `engine/rewrite/rewrite_generation.py` — `RewriteGenerationWorkflow.run()`.
- `engine/detection/detection_workflow.py` — similar split pattern.

## Proposal

Extract a shared helper, e.g.:

```python
def run_on_entity_rows(
    df: pd.DataFrame,
    *,
    has_entities_fn: Callable[[Any], bool],
    passthrefaults: dict[str, Any],
    run_fn: Callable[[pd.DataFrame], WorkflowRunResult],
) -> WorkflowRunResult:
```

Handles partitioning, ordering, recombination, and `attrs` propagation in one place.

## Acceptance criteria
- [ ] Shared helper in `engine/utils.py` or similar.
- [ ] `generate_map_only` and `RewriteGenerationWorkflow.run()` refactored to use it.
- [ ] No behavior change — existing tests pass as-is.

### PR DESCRIPTION
**Pull request by andreatnvidia**
## Summary

- Extract the duplicated split/reorder/recombine DataFrame pattern into a shared `engine/row_partitioning.py` module with `split_rows()` and `merge_and_reorder()`
- Refactor `rewrite_workflow.py` and `llm_replace_workflow.py` to use the shared helpers instead of inline implementations
- Net reduction of ~46 lines of duplicated logic

Closes #60

## Test plan

- [x] 10 new unit tests for `row_partitioning.py` (partitioning, edge cases, roundtrip)
- [x] All 495 existing tests pass with no changes
- [x] Lint clean (`ruff check`)

### CODE REVIEW

File: src/anonymizer/engine/row_partitioning.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+"""Shared helpers for partitioning DataFrames by a row predicate and recombining them."""
+
+from __future__ import annotations
+
+from collections.abc import Callable
+from typing import Any
+
+import pandas as pd
+
+_ROW_ORDER_COL = "_anonymizer_row_order"
```
Comment (lipikaramaswamy) [inline src/anonymizer/engine/row_partitioning.py]: nit: Now that this is the single source of truth, consider dropping the leading underscore so tests can import it:

```
ROW_ORDER_COL = "_anonymizer_row_order"
```
`test_rewrite_workflow.py` (8 places) and `test_llm_replace_workflow.py` (3 places) hard-code "_anonymizer_row_order" in mock fixtures. Having them import the constant prevents silent drift if the value ever changes.

File: src/anonymizer/engine/row_partitioning.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    can restore the original ordering after the two partitions are processed
+    independently.
+    """
+    working = df.copy()
+    working[_ROW_ORDER_COL] = range(len(working))
+    mask = working[column].apply(predicate)
+    return working[mask].copy(), working[~mask].copy()
+
+
+def merge_and_reorder(
```
Comment (lipikaramaswamy) [inline src/anonymizer/engine/row_partitioning.py:34]: nit: `pd.concat([])` raises ValueError if no positional args are passed. Current callers always pass at least one part, but a guard like below would make the contract explicit. 

```
if not parts:
    raise ValueError("merge_and_reorder requires at least one partition")
```

File: src/anonymizer/engine/replace/llm_replace_workflow.py [hunk 3]
```diff
@@ -103,19 +95,10 @@ def generate_map_only(
...
-            .sort_values("_anonymizer_row_order")
-            .drop(columns="_anonymizer_row_order")
-            .reset_index(drop=True)
+        combined = merge_and_reorder(
+            output_df, passthrough_rows, attrs={**run_result.dataframe.attrs, **dataframe.attrs}
         )
-        # Carry pipeline metadata (e.g. original_text_column) through to downstream steps.
-        # pandas .attrs is experimental and not preserved through merge/concat/groupby,
-        # but is preserved through copy which is all we do here.
-        # TODO: consider wrapping df + metadata in a typed container.
```
Comment (lipikaramaswamy) [inline src/anonymizer/engine/replace/llm_replace_workflow.py:116]: Just noting it's ok to drop this TODO even though it's not addressed because we have issue https://github.com/NVIDIA-NeMo/Anonymizer/issues/4

File: tests/engine/test_row_partitioning.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    assert combined.attrs == attrs
+
+
+def test_merge_and_reorder_single_partition() -> None:
+    df = pd.DataFrame({"val": [1, 2, 3]})
+    matching, _ = split_rows(df, column="val", predicate=bool)
+    combined = merge_and_reorder(matching, attrs={})
+    assert list(combined["val"]) == [1, 2, 3]
+
+
```
Comment (lipikaramaswamy) [inline tests/engine/test_row_partitioning.py:91]:
```suggestion

def test_merge_and_reorder_raises_on_empty_parts() -> None:
    with pytest.raises(ValueError, match="at least one partition"):
        merge_and_reorder(attrs={})
```

consider adding to check the guard I suggested above

File: tests/engine/test_row_partitioning.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+def test_roundtrip_preserves_columns_added_after_split() -> None:
+    df = pd.DataFrame({"val": [1, 0, 3]})
+    matching, non_matching = split_rows(df, column="val", predicate=bool)
+
+    matching["extra"] = "processed"
+    non_matching["extra"] = "skipped"
+
+    combined = merge_and_reorder(matching, non_matching, attrs={})
+    assert list(combined["extra"]) == ["processed", "skipped", "processed"]
```
Comment (lipikaramaswamy) [inline tests/engine/test_row_partitioning.py:101]: Worth also asserting on the column the split was based on.
```
assert list(combined["val"]) == [1, 0, 3]
```

Currently if `merge_and_reorder` scrambled the base data but got `"extra"` right by coincidence, this test would still pass.

**Review by lipikaramaswamy — APPROVED** [review]
Looks good, just a few nits and suggestions. e2e test ran without issues.

File: src/anonymizer/engine/row_partitioning.py (hunk 1 above)
Comment (andreatnvidia) [reply src/anonymizer/engine/row_partitioning.py]: Done - tests import the constant now.

File: src/anonymizer/engine/row_partitioning.py (hunk 2 above)
Comment (andreatnvidia) [reply src/anonymizer/engine/row_partitioning.py:34]: Added. Better than a cryptic `pd.concat` error.

File: src/anonymizer/engine/replace/llm_replace_workflow.py (hunk 3 above)
Comment (andreatnvidia) [reply src/anonymizer/engine/replace/llm_replace_workflow.py:116]: Yep, leaving as-is since #4 tracks it.

File: tests/engine/test_row_partitioning.py (hunk 4 above)
Comment (andreatnvidia) [reply tests/engine/test_row_partitioning.py:91]: Added.

File: tests/engine/test_row_partitioning.py (hunk 5 above)
Comment (andreatnvidia) [reply tests/engine/test_row_partitioning.py:101]: Good catch, added.
